lib/eosknowledgeprivate/ekn-media-bin.c

```diff
@@ -0,0 +1,1606 @@
+  gst_bus_add_watch (self->bus, ekn_media_bin_bus_watch, self);
+}
+
+/********************************* Public API *********************************/
+
+/**
+ * ekn_media_bin_new:
+ *
+ * Returns a new #EknMediaBin
+ *
+ */
+GtkWidget *
+ekn_media_bin_new ()
+{
+  return (GtkWidget*) g_object_new (EKN_TYPE_MEDIA_BIN, NULL);
+}
+
+/* The following macros are used to define generic getters and setters */
+
+#define EMB_DEFINE_GETTER(type, prop, retval) \
+type \
+ekn_media_bin_get_##prop (EknMediaBin *self) \
+{ \
+  g_return_val_if_fail (EKN_IS_MEDIA_BIN (self), retval); \
+  return self->prop; \
+}
+
+/* The last argument is for custom code that will be added just before calling
+ * g_object_notify_by_pspec()
+ *
+ * You have access to EknMediaBin *self and the property defined by prop
+ */
+#define EMB_DEFINE_SETTER(type, prop, PROP, code) \
+void \
+ekn_media_bin_set_##prop (EknMediaBin *self, type prop) \
+{ \
+  g_return_if_fail (EKN_IS_MEDIA_BIN (self)); \
+  if (self->prop != prop) \
+    { \
+      self->prop = prop; \
+      code \
+      g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_##PROP]); \
+    } \
+}
+
+#define EMB_DEFINE_SETTER_STRING(prop, PROP, code) \
+void \
+ekn_media_bin_set_##prop (EknMediaBin *self, const gchar *prop) \
+{ \
+  g_return_if_fail (EKN_IS_MEDIA_BIN (self)); \
+  if (g_strcmp0 (self->prop, prop)) \
+    { \
+      g_free (self->prop); \
+      self->prop = g_strdup (prop); \
+      code \
+      g_object_notify_by_pspec (G_OBJECT (self), properties[PROP_##PROP]); \
+    } \
+}
+
+/**
+ * ekn_media_bin_get_uri:
+ * @self: a #EknMediaBin
+ *
+ * Return the media URI
+ */
+EMB_DEFINE_GETTER (const gchar *, uri, NULL)
+
+/**
+ * ekn_media_bin_set_uri:
+ * @self: a #EknMediaBin
+ * @uri:
+ *
+ * Sets the media URI to play
+ */
+EMB_DEFINE_SETTER_STRING (uri, URI,
+  /* Make playbin show the first video frame if there is an URI
+   * and the widget is realized.
+   */
+   ekn_media_bin_set_state_ready (self);
+)
+
+/**
+ * ekn_media_bin_get_autohide_timeout:
+ * @self: a #EknMediaBin
+ *
+ * Returns control's auto hide timeout in seconds.
+ */
+EMB_DEFINE_GETTER (gint, autohide_timeout, 0)
+
+/**
+ * ekn_media_bin_set_autohide_timeout:
+ * @self: a #EknMediaBin
+ * @autohide_timeout: A timeout in seconds
+ *
+ * Sets the timeout to auto hide controls
+ */
+EMB_DEFINE_SETTER (gint, autohide_timeout, AUTOHIDE_TIMEOUT,)
+
+/**
+ * ekn_media_bin_get_fullscreen:
+ * @self: a #EknMediaBin
+ *
+ * Returns whether video is fullscreen or not
+ */
+EMB_DEFINE_GETTER (gboolean, fullscreen, FALSE)
+
+/**
+ * ekn_media_bin_set_fullscreen:
+ * @self: a #EknMediaBin
+ * @fullscreen:
+ *
+ * Sets whether to show the video in fullscreen mode or not
+ */
+EMB_DEFINE_SETTER (gboolean, fullscreen, FULLSCREEN,
+  /* If there is no video sink, delay fullscreen until realize event */
+  if (self->video_sink)
+    ekn_media_bin_fullscreen_apply (self, fullscreen);
```

Comment:
Obviously this is working, but just out of curiosity, how? Shouldn't everything after the comma in this line be treated like a separate argument to the macro?